src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
+        self.set_hover_split_offset(Vector2::new(INFINITE, INFINITE));
+        self.set_hover_split_rotation(RIGHT_ANGLE);
+    }
+
+    /// Convert the given global coordinates to the shape local coordinate system.
+    fn to_shape_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = point - self.display_object().global_position().xy();
+        UnitComplex::new(-base_rotation) * local_unrotated
+    }
+
+    /// Convert the given shape local coordinate to the global coordinates system.
+    fn from_shape_to_global_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = UnitComplex::new(base_rotation) * point;
+        local_unrotated + self.display_object().global_position().xy()
+    }
+
+    /// Check whether the given shape local coordinate is within the bounds of the shape.
+    fn contains_local(&self, point:Vector2<f32>) -> bool {
+        let size = self.sprite().size.get();
+        if point.x < -size.x / 2.0 || point.y < -size.y / 2.0 {
+            return false
+        }
+        point.x < size.x / 2.0 && point.y < size.y / 2.0
+    }
+
+    /// Check whether the given global coordinate is within the bounds of the shape.
+    fn contains(&self, point:Vector2<f32>) -> bool {
+        self.contains_local(self.to_shape_coordinate_system(point))
+    }
+}
+
+
+
+// ========================
+// === Edge Shape Trait ===
+// ========================
+
+/// Indicates which area should be recolored.
+#[derive(Clone,Copy,Debug,Eq,PartialEq)]
+enum Area {
+    Above,
+    Below,
+}
+
+/// Holds the data required to split a shape into two parts.
+///
+/// The `area` indicates which side of the split will receive special coloring.
+#[derive(Clone,Copy,Debug)]
+struct SplitData {
+    split_position : Vector2<f32>,
+    area           : Area,
+    cut_angle      : f32
+}
+
+impl SplitData {
+    fn new(split_position:Vector2<f32>, area: Area, cut_angle:f32) -> Self {
+        SplitData {split_position,area,cut_angle}
+    }
+}
+
+/// `SnapData` is used as the result value of snapping operations on `MultiShapes`. It holds the
+/// position and shape that a hover position was snapped to, that is the shape and the position
+/// on the shape that was closest to the hover
+struct SnapData<'a> {
+    position     : Vector2<f32>,
+    target_shape : &'a dyn EdgeShape,
+}
+
+impl<'a> SnapData<'a> {
+    fn new(position: Vector2<f32>, target_shape: &'a dyn EdgeShape) -> Self {
+        SnapData{position,target_shape}
+    }
+}
+
+/// The MultiShape trait allows operations on a collection of `EdgeShape`.
+trait MultiShape {
+    /// Return the `ShapeViewEvents` of all sub-shapes.
+    fn events(&self) -> Vec<ShapeViewEvents>;
+    /// Return references to all `EdgeShape`s in this MultiShape.
+    fn edge_shape_views(&self) -> Vec<&dyn EdgeShape>;
```

Comment:
Instead of returning references here, it might be better to use `Id`s to avoid lifetime issues.